snuba/datasets/factory.py

```diff
@@ -17,7 +20,7 @@ class InvalidDatasetError(Exception):
     """Exception raised on invalid dataset access."""
 
 
-def get_dataset(name):
+def get_dataset(name: str) -> Dataset:
```

Comment:
What would you think about using an enum to identify datasets? I have a hard time developing a strong opinion on this:
Pros of the enum:
- harder to introduce typos
- cleaner to validate input parameters (when they represent a dataset name)

Cons of the enum:
- requires a code change in the factories module (core Snuba module) every time a new dataset is added. Which is totally ok now, but in a world where dataset are more driven by configuration this becomes a pain.
- It makes it factually impossible to separate Snuba (as a storage systems) from its datasets to make it more Sentry agnostic like we did for CDC.

Opinions ?

Reply:
I think the main benefit of the enum here is that it'd be harder to introduce typos strictly because of the ability to type check the enum values? I think the runtime error scenarios would be more or less the same as today.

My ideal design is one where dataset definition happens purely in configuration (similar to CDC, like you mentioned), where there'd be very loose coupling, preferably to the point where no component refers to another component by name. I don't think an enum makes much sense in that case, since there's really no reason to be referring to specific datasets in code — and correspondingly, no reason to improve the ergonomics around type checking them.

We're obviously a long way away from that goal, and I don't have a strong opinion one way or another. I think that type checking dataset classes like in https://github.com/getsentry/snuba/pull/485#discussion_r328804602 has met my needs, and I'm not sure how much advantage an enum would have over that approach?